docs/tutorials/index.md

```diff
@@ -0,0 +1,186 @@
+same code that is used to process live data -- and to produce deterministic, consistent results,
+regardless.
+
+It can also be crucial to pay attention to the order in which events occurred, rather than the order
+in which they are delivered for processing, and to be able to reason about when a set of events is
+(or should be) complete. For example, consider the set of events involved in an e-commerce
+transaction, or financial trade.
+
+These requirements for timely stream processing can be met by using event time timestamps that are
+recorded in the data stream, rather than using the clocks of the machines processing the data.
+
+{% top %}
+
+## Stateful Stream Processing
+
+Flink's operations can be stateful. This means that how one event is handled can depend on the
+accumulated effect of all the events that came before it. State may be used for something simple,
+such as counting events per minute to display on a dashboard, or for something more complex, such as
+computing features for a fraud detection model.
+
+A Flink application is run in parallel on a distributed cluster. The various parallel instances of a
+given operator will execute independently, in separate threads, and in general will be running on
+different machines.
+
+The set of parallel instances of a stateful operator is effectively a sharded key-value store. Each
+parallel instance is responsible for handling events for a specific group of keys, and the state for
+those keys is kept locally.
+
+The diagram below shows a job running with a parallelism of two across the first three operators in
+the job graph, terminating in a sink that has a parallelism of one. The third operator is stateful,
+and you can see that a fully connected network shuffle is occurring between the second and third
+operators. This is being done to partition the stream by some key, so that all of the events that
+need to be processed together, will be.
+
+<img src="{{ site.baseurl }}/fig/parallel-job.png" alt="State is sharded" class="offset" width="65%" />
+
+State is always accessed locally, which helps Flink applications achieve high throughput and
+low-latency. You can choose to keep state on the JVM heap, or if it is too large, in efficiently
+organized on-disk data structures. 
+
+<img src="{{ site.baseurl }}/fig/local-state.png" alt="State is local" class="offset" width="90%" />
+
+{% top %}
+
+## Fault Tolerance via State Snapshots
+
+Flink is able to provide fault-tolerant, exactly-once semantics through a combination of state
+snapshots and stream replay. These snapshots capture the entire state of the distributed pipeline,
+recording offsets into the input queues as well as the state throughout the job graph that has
+resulted from having ingested the data up to that point. When a failure occurs, the sources are
+rewound, the state is restored, and processing is resumed. As depicted above, these state snapshots
+are captured asynchronously, without impeding the ongoing processing.
```

Comment:
"Understatement of the year" ;)
But as for the training and as an introduction, it is fine to keep away the details here.

Reply:
Lol.